src/Microsoft.HttpRepl/Program.cs

```diff
@@ -23,9 +24,8 @@ public static async Task Main(string[] args)
         }
 
         public async Task Start(string[] args, IConsoleManager console)
-        { 
-            IFileSystem fileSystem = new RealFileSystem();
-            HttpState state = CreateHttpState(fileSystem);
+        {           
+            HttpState state = CreateHttpState(out IFileSystem fileSystem, out IPreferences preferences);
```

Comment:
Hm, the purist in me is saying that this is fragmenting the composition root - the dependencies are being initialized in a separate place from the rest of the dependency graph (here).  That fragmentation was actually introduced in the last Preferences change, when `CreateHttpState` was created.

What do you think about formalizing composition more?  It's really not a responsibility of the Start method - what if we had an explicit root like `ComposeDependencies(IConsole console, out Shell shell, out IPreferences prefs)` (to start with - it may evolve over time, though hopefully not - console is already kind of a hack to enable the integration tests, but the rest of the composition for the app should be internal to the composition root even if types/dependencies change), and that would take the CreateHttpState and dispatcher setup all into one place, leaving Start as something like (pseudo):

```
Start(string[] args, IConsole console) 
{
    ComposeDependencies(console, out Shell shell, out IPreferences prefs);
    if (Console.IsOutputRedirected && !console.AllowOutputRedirection)
    {
        Reporter.Error.WriteLine(Resources.Strings.Error_OutputRedirected.SetColor(prefs.GetValue(WellKnownPreference.ErrorColor)));
        return;
    }

    CancellationTokenSource source = new CancellationTokenSource();
~~    Shell shell = new Shell(dispatcher, consoleManager: console);~~ // this line no longer needed
    shell.ShellState.ConsoleManager.AddBreakHandler(() => source.Cancel());
    /* etc... */
}
```

Reply:
Makes sense to me. I gave it a shot in c430a00, let me know what you think.

Reply:
I like it 👍